fix(fmgc): pre-selected speeds

MisterChocker opened the pull request:
<!-- Signed-off-by: Leon Beeser <leon.beeser@yahoo.de> -->

<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->
Fixes #4531

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
- Added missing fmgc change to changelog
- Re-add pre-selected speeds handling
- Added FMA simvar support
- Added support for Mach
- Improved performance page input handling
- Added FCU handler to override FCU speed value
- Added basic logic for selected and preselected modes on perf pages

<!-- ## Screenshots (if necessary) -->
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make your best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->
![image](https://user-images.githubusercontent.com/1857322/116820923-c0db7d80-ab77-11eb-924e-38cc690d80b9.png)
![image](https://user-images.githubusercontent.com/1857322/116820928-c5a03180-ab77-11eb-8131-12a1d9562c30.png)


<!-- ## References -->
<!-- You should be making changes based on some kind of a reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->
For other contributors:
The FMS now assigns pre-sections to a dedicated variable inside FMS.

When changing speeds all new values are written to their own variable AND when the value set is the pre-sel-speed for the next phase it will write the new value also to the corresponding simvar (either speed or mach) and ensures only one of these two are defined.

When switching flight phases, the FMS will send a command to the FCU which then sets the requested selected speed/mach value.

To ensure this process is successful, the FMS will wait ~200ms before clearing and overriding the pre-selected-speed/mach simvars.

Known issue: Performance page needs some time to display the correct selected speed.

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub): DerL30N#3751

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->
Enter a pre-selected speed or mach into the corresponding performance page.
In the flight phase before the flight phase with mach/speed pre selected, the FMA indicates that mach or speed has been preselected.
When transitioning into corresponding flight phase -> ac targets pre-selected speed/mach (in selected mode)

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

frankkopp commented:
Did a quick test on this although FCU part not yet implemented it seems the pre-selected speeds work.
It kept CLB speed to the selected value and also CRZ speed when reaching CRZ alt. 
FCU obviously did not update and stayed at 100.

frankkopp commented:
I know this is not ready yet. Did a little testing and wanted to share this PSD issue:
![image](https://user-images.githubusercontent.com/16833201/116827633-4bcd6f80-ab9a-11eb-8dc2-955665cfd5d1.png)

Otherwise everything seems to work but I will wait with further testing until ready to test.

lukecologne commented:
> 
> 
> I know this is not ready yet. Still did a little testing and wanted to share this PSD issue:
> 
> Otherwise everything seems to work but I will wait with further testing until ready to test.

This will be fixed with #4695

tracernz commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FCU/A320_Neo_FCU.js:
```diff
@@ -432,6 +432,24 @@ class A320_Neo_FCU_Speed extends A320_Neo_FCU_Component {
+            this.isTargetManaged = false;
+            this.isMachActive = true;
+            this.selectedValue = SimVar.GetSimVarValue("L:A32NX_MachPreselVal", "mach");
+            SimVar.SetSimVarValue("K:AP_MANAGED_SPEED_IN_MACH_ON", "number", 1);
```
Maybe better to factor these two into a function like the existing ones, like `onPreSelSpeed(mach)`, so a) you can de-duplicate most of the code, and b) to keep the flow simple and more readable in this function.

tracernz commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:
```diff
@@ -362,14 +368,17 @@ class FMCMainDisplay extends BaseAirliners {
-                if (SimVar.GetSimVarValue("L:A32NX_GOAROUND_PASSED", "bool") === 1) {
-                    preSelectedClbSpeed = this.computedVgd;
+                /** Activate pre selected speed/mach */
+                // this.activatePreSelSpeedMach(_lastFlightPhase === FmgcFlightPhases.GOAROUND ? ... : this.preSelectedClbSpeed);
```
Is this comment still needed? Looks like old logic you replaced.

MisterChocker replied:
Yes the comment is about the old logic, it would've preselected GD speed during GO AROUND.

I haven't yet checked this against the FCOM/manual.

tracernz commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FCU/A320_Neo_FCU.js:
```diff
@@ -432,6 +432,24 @@ class A320_Neo_FCU_Speed extends A320_Neo_FCU_Component {
             this.onSwitchSpeedMach();
+        } else if (_event === "USE_PRE_SEL_SPEED") {
+            clearTimeout(this._resetSelectionTimeout);
+            SimVar.SetSimVarValue("K:SPEED_SLOT_INDEX_SET", "number", 1);
```
Maybe this should be done after the `selectedValue` is set. In practice I don't think it will matter.

MisterChocker replied:
I literally copy and pasted this code from the pull() event of the fcu. Sure the order variables are set can be changed, bit I honestly don't see why xD

tracernz commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:
```diff
@@ -2477,9 +2530,30 @@ class FMCMainDisplay extends BaseAirliners {
 
     //TODO: fix this functionality
     trySetPreSelectedClimbSpeed(s) {
+        const isNextPhase = this.currentFlightPhase === FmgcFlightPhases.TAKEOFF;
```
According to FCOM the preselected speed cannot be modified on the MCDU once the phase is active, only by rotating the FCU knob.
<img width="705" alt="image" src="https://user-images.githubusercontent.com/9995998/116977907-1b5f1000-ad17-11eb-80aa-03ce8811293b.png">

MisterChocker replied:
Yes and this is true, isNextPhase just ensures that the pre-sel speed I am about to modify is is the next one in order to ensure correct population of the simvar.

That variable has nothing todo whether a value can be changed or not.

tracernz commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:
```diff
@@ -2488,9 +2562,30 @@ class FMCMainDisplay extends BaseAirliners {
 
     //TODO: fix this functionality
     trySetPreSelectedCruiseSpeed(s) {
+        const isNextPhase = this.currentFlightPhase === FmgcFlightPhases.CLIMB;
```
No entry once in cruise.
<img width="697" alt="image" src="https://user-images.githubusercontent.com/9995998/116978022-43e70a00-ad17-11eb-8384-f6e5edad56b0.png">
that note...

tracernz commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:
```diff
@@ -2499,9 +2594,30 @@ class FMCMainDisplay extends BaseAirliners {
 
     //TODO: fix this functionality
     trySetPreSelectedDescentSpeed(s) {
+        const isNextPhase = this.currentFlightPhase === FmgcFlightPhases.CRUISE;
```
Doesn't look like this should be editable according to FCOM.
<img width="716" alt="image" src="https://user-images.githubusercontent.com/9995998/116978230-8a3c6900-ad17-11eb-8778-35566f4f464f.png">

beheh replied:
This is possible in FMS release 2.

beheh replied:
Disregard, I was talking about changing the managed speed in 3L, not the selected speed in 4L

tracernz reviewed (commented):
Really nice work! I've missed this functionality a few times before so it's nice to see it.

I have not checked what the Honeywell manual says so it may contradict the FCOM references I used above. Apologies if that's the case, time is not on my side today. :(

lukecologne reviewed (commented):
The Simvars you are using for the presel values are being initialized in the PFD as -1. This should be removed, and handled in your code:
https://github.com/flybywiresim/a32nx/blob/3d8460b42da97b70a8ecd2f1258d1fb4f8d262c3/src/instruments/src/PFD/index.jsx#L46-L47

tracernz reviewed (approved):
Great work 🚀

frankkopp commented:
Category: QA Tester
Name: Cdr_Maverick#6475
Date of testing: 9.05.2021
Version of the sim: 1.15.10.0
PR Tested: #4555
Tier of testing: 2

Changes to observe:
Enter a pre-selected speed or mach into the corresponding performance page.
In the flight phase before the flight phase with mach/speed pre selected,
the FMA indicates that mach or speed has been preselected. When transitioning
into corresponding flight phase
-> ac targets pre-selected speed/mach (in selected mode)

Testing technique: Tryout starting on runway

Steps made to test the Pull Request:
1)	Made several shorts flight, start on runway, loaded Simbrief flight.
2)	PERF: set selected speed for CLB to 220, set sel speed for CRZ to 320 , set sel speed for DES to 275
3)	Take-off, no AP until after LVR to CLB – checked Speed Mode = Managed (no overlapping text any more)
4)	When flight phase changes to CLB Speed Mode changed to Selected Speed at 220
5)	PFD now shows SELECTED (or MACH) SEL 320
6)	Settings Managed Speed >10.000ft - PERF page changes to MANAGED, speed goes to *290 as per PERF page
7)	At cruise alt flight phase change to CRZ and Speed Mode changes to Selected Speed at 320
8)  PERF page shows Selected 320 / PFD shows SPEED SEL 275
9)	Descending to 8000ft in managed descend
10) Flight phase change to DES and Speed Mode stays in Selected Speed at 275
11) PERF page shows Selected 275
12) PFD not showing SPEED SEL any longer
13) Second test with Mach values

Overall Rating: Good

Conclusions of the testing: Works as expected.
When in CLB/CRZ/DES mode with Selected Speed it is still possible to change value in
PERF page. But has no effect on actual Selected Speed. Is this correct?

MisterChocker commented:
> When in CLB/CRZ/DES mode with Selected Speed it is still possible to change value in
> PERF page.

No, you should not be able to change/modify the preselected speed for the same flight phase since it then should only display the selected fcu speed.

I'll look into it.

frankkopp commented:
Re-tested:
Value on PERFT page of current phase can't be changed any longer. 

Issues:
- Slight delay of a few seconds betwenn phase change and PERF page change from Managed to Selected - intended?
- UI issue - text not aligned to the left
![image](https://user-images.githubusercontent.com/16833201/117580713-92225180-b0f9-11eb-8d67-057b4142a50d.png)
:

MisterChocker commented:
> * UI issue - text not aligned to the left

I've adjusted the rows 2/3/4L to follow the FCOM.

> * Slight delay of a few seconds betwenn phase change and PERF page change from Managed to Selected - intended?

I've already added this to the known issues. Will fix in the future (aka performance page rewrite)

frankkopp commented:
> FCOM
You are of course right - double checked it in the FCOM (had to get one first - do you have one for the neo?). 
Weired format on the PERF page :)

TheSquirtingPantin commented:
Category: QA Tester
Name: Discord User Rémi#8871
Date of testing: Current date 09/05/2021
Version of the sim: V 1.15.8.0
PR Tested: #4555
Tier of testing: 2
Changes to observe: Pre-Selected Speed function

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. C/D at LFBO
2. Full Flight
3. Observe logic of MCDU PERF page with Pre Selected Speed

Flight notes (When done a full flight): LFBO/32L DCT LACOU UT24 SECHE UL127 SAU DCT SOMOS DCT LMB DCT LFBO/32L

Media:
![Infinite number](https://user-images.githubusercontent.com/19536495/117593383-2643e700-b101-11eb-96b8-be970fe5644a.PNG)

Issues (If Any):
1) During descent when using selected speed (with FCU control knob) in "Knots" and changing to "Mach", the number generated on the MCDU PERF page is displayed with no round-out (see picture). 

Overall Rating: To be Improved
Conclusions of the testing: Overall logic looks good to me. Number rounding issue when switching from Mach to Knots.

MisterChocker commented:
@TheSquirtingPantin Oh you are right, I've forgot to test with mach selection on the FCU.

frankkopp commented:
An additional observation of this last Mach/Knots issue is that if you switch back and forth between Mach/Knots the value gets smaller every time. Not a realistic use case but still something that should not happen.

dragold commented:
Category: QA Tester Trial
Name: Discord User Mr.Blank#2638
Date of testing: Current date 10/05/2021
Version of the sim: 1.15.10.0
PR Tested: #4555
Tier of testing: 2
Changes to observe: Pre-Selected Speed function
Testing technique: Full flight try out

1 SOP flight
1.1 PFD and FMGC observations
2 Non- SOP Flight  (Quickstart)
2.1 PFD and FMGC observations

Flight notes (When done a full flight): ROUTE ID: DEFRTE
EDHI/23 HAM1V HAM DCT BOGMU DCT ABGEB DCT GIKOG DCT LEGPI DCT DHE
DCT NOSNO DCT RIBSO DCT EDHI/23

Issues (If Any): NIL

Remark: When changing from its to M number while CRZ with pre-entered selected the number change(MCDU PERF page) is no round-out 

Overall Rating: Good, but can be improved 
Conclusions of the testing: The observations whereas expected however the item referred to in the remark need to be reviewed.. Good luck, best wishes

MisterChocker commented:
I've fixed two things:
- Managed speed not being populated
- Selected speed not being rounded

frankkopp commented:
Category: QA Tester
Name: Cdr_Maverick#6475
Date of testing: Current date 15.5.2021
Version of the sim: 1.15.10
PR Tested: #4555
Tier of testing: 1/2/3

Changes to observe:
Enter a pre-selected speed or mach into the corresponding performance page. In the flight phase
before the flight phase with mach/speed pre selected, the FMA indicates that mach or speed has
been preselected. When transitioning into corresponding flight phase -> ac targets pre-selected
speed/mach (in selected mode)

Testing technique: Testing KTS/Mach issue and rounding
Steps you've made to test the Pull Request:
1. Started on runway for CLB, CRZ and DES
2. Selected Speed works as tested before
3. Switching to mach and back does not show rounding issue any more - value always full integer
4. Switching back and forth between KTS/Mach still changes the selected speed (likely through rounding error)

Issues (If Any): Minor issue when switching between SPD/Mach it changes SPD value and also value on Selected Speed Perf page

Overall Rating: Mostly Good (switching between SPD and Mach is not a normal use case but could be improved anyway)
Conclusions of the testing: LGTM (and maybe fix the minor issue in another PR - this feature is too important imho to not merge it because of this)

dragold commented:
Category: QA Tester Trial
Name: Discord User Mr.Blank#2638
Date of testing: Current date 16/05/2021
Version of the sim: 1.15.10.0
PR Tested: #4555
Tier of testing: 2
Changes to observe: Pre-Selected Speed function
Testing technique: Full flight try out

1 SOP flight
1.1 PFD and FMGC observations
2 Non- SOP Flight (Quickstart)
2.1 PFD and FMGC observations

Flight notes (When done a full flight): ROUTE ID: DEFRTE
EDHI/23 HAM1V HAM DCT BOGMU DCT ABGEB DCT GIKOG DCT LEGPI DCT DHE
DCT NOSNO DCT RIBSO DCT EDHI/23

Issues (If Any): 
1- speed value between the PFD and FMCDU is not equal
![image](https://user-images.githubusercontent.com/47900069/118379651-8eed0100-b5e4-11eb-88d1-bc3e895e1603.png)


Remark:- when the speed is selected above High-Speed Protection in some cases the speed above High-Speed Protection without activating it (more test/s needs to be done to try to reproduce it) 


Overall Rating: Can be improved
Conclusions of the testing: Some observations were no as expected, error log (taken at time of bug found) attached to the report Good luck, best wishes...
[error log #4555.txt](https://github.com/flybywiresim/a32nx/files/6488363/error.log.4555.txt)

aguther commented:
In regards to high speed protection you need to know the following:

- it's only active above VMO/MMO *not* in red band with flaps out
- when AP is engaged it activates quite late (i.e. 3 seconds above 365 kn)

Now with the latest changes on master in OP CLB/DES it will hold at max VMAX or VLS as Minimum. But that's simply Limitation of the target Speed no high speed protection

MisterChocker commented:
> 1- speed value between the PFD and FMCDU is not equal

The update rate of the performance page is lower and therefore it takes some time before the value has updated.
The speed value on the PFD and MCDU use the exact same unit conversion and rounding method.
